apps/meeting/src/providers/VideoGridProvider/state.tsx

```diff
@@ -0,0 +1,319 @@
+    for (let i = 1; i < zoomedLevels.length - 1; i += 1) {
+      if (zoomedLevels[i] === size) {
+        zoomedLevelIndex = i;
+      } else if (zoomedLevels[i] < size) {
+        zoomedLevelIndex = i + 1;
+      }
+    }
+    zoomedLevel = zoomedLevels[zoomedLevelIndex];
+  };
+
+  const calculateVideoSourcesToBeRendered = (): string[] => {
+    const left = (currentPage - 1) * zoomedLevel;
+    const right = currentPage * zoomedLevel;
+    return availableVideoSources.slice(left, right);
+  };
+
+  const updateDownlinkPreferences = (): void => {
+    const videoSourcesToBeRendered = calculateVideoSourcesToBeRendered();
+    const videoPreferences = VideoPreferences.prepare();
+    const thresholds = 8;
+    let targetDisplaySize: TargetDisplaySize;
+
+    if (videoSourcesToBeRendered.length < thresholds) {
+      targetDisplaySize = TargetDisplaySize.High;
+    } else {
+      targetDisplaySize = TargetDisplaySize.Low;
+    }
+
+    for (const attendeeId of videoSourcesToBeRendered) {
+      videoPreferences.add(
+        new VideoPreference(attendeeId, 1, targetDisplaySize)
+      );
+    }
+    priorityBasedPolicy.chooseRemoteVideoSources(videoPreferences.build());
+  };
+
+  switch (type) {
+    case VideoGridAction.ResetVideoGridState: {
+      return initialState;
+    }
+    case VideoGridAction.UpdateVideoSources: {
+      const { videoSources } = payload;
+
+      // Reset the `videoEnabled` of all attendees
+      for (const attendee of Object.values(attendees)) {
+        attendee.videoEnabled = false;
+      }
+
+      // Update the `videoEnabled` based on `videoSources`
+      for (const videoSource of videoSources) {
+        const { attendee } = videoSource;
+        const { attendeeId } = attendee;
+
+        // To do: check if this triggers only by content sharing
+        if (!(attendeeId in attendees)) {
+          attendees[attendeeId] = { attendeeId } as AttendeeStatus;
+        }
+        attendees[attendeeId].videoEnabled = true;
+      }
+
+      // Populate the `availableVideoSources` based on the order of attendees
+      const newAvailableVideoSources: string[] = [];
+      for (const attendee of Object.values(attendees)) {
+        if (attendee.videoEnabled) {
+          newAvailableVideoSources.push(attendee.attendeeId);
+        }
+      }
+
+      availableVideoSources = newAvailableVideoSources;
+
+      if (availableVideoSources.length === 0) {
+        currentPage = 1;
+        isZoomed = false;
+        zoomedLevel = 0;
+        zoomedLevelIndex = 0;
+      } else {
+        if (!isZoomed) {
+          // If not manually zoomed in, change the zoomLevel accordingly
+          updateZoomStatus();
+        }
+        updateDownlinkPreferences();
+      }
+
+      return {
+        attendees,
+        availableVideoSources,
+        currentPage,
+        isZoomed,
+        zoomedLevel,
+        zoomedLevelIndex,
+      };
+    }
+    case VideoGridAction.UpdateRoster: {
```

Comment:
Could we rename this UpdateRoster to something else? It does not updateRoster just set video flag.

Reply:
Updated to `UpdateAttendeeStates`